lib/galaxy/tools/parameters/basic.py

```diff
@@ -2029,6 +2031,30 @@ def do_validate(v):
                 raise ValueError("At most %d datasets are required for %s" % (self.max, self.name))
 
 
+def src_id_to_item(
+    sa_session: "Session", value: Dict[str, Any], security: "IdEncodingHelper"
+) -> Union[
+    DatasetCollectionElement,
+    HistoryDatasetAssociation,
+    HistoryDatasetCollectionAssociation,
+    LibraryDatasetDatasetAssociation,
+]:
+    src_to_class = {
+        "hda": HistoryDatasetAssociation,
+        "ldda": LibraryDatasetDatasetAssociation,
+        "dce": DatasetCollectionElement,
+        "hdca": HistoryDatasetCollectionAssociation,
+    }
+    id_value = value["id"]
+    decoded_id = id_value if isinstance(id_value, int) else security.decode_id(id_value)
+    try:
+        item = sa_session.query(src_to_class[value["src"]]).get(decoded_id)
```

Comment:
```suggestion
        item = sa_session.get(src_to_class[value["src"]], decoded_id)
```

that should be the new sqlalchemy 2.0 compatible syntax